### ISSUE TITLE
DateIterator returning unexpected recurrence values with advanceTo()

### ISSUE THREAD

**Issue by agiannone** [mangstadt/biweekly#69]
Hi, I am getting unexpected recurrence values from the DateIterator. See the test below:

```
Instant awardedOn = Instant.from( format.parse( "2017-04-04T01:19:24.162Z" ) );
Instant periodStart = Instant.from( format.parse( "2016-11-27T00:00:00.000Z" ) );

Recurrence rec = new Recurrence.Builder( Frequency.HOURLY ).interval( 1 ).build();
DateIterator iter = rec.getDateIterator( Date.from( periodStart ), TimeUtils.TIME_ZONE );
iter.advanceTo( Date.from( awardedOn ) );

Instant recurrence = null;
if( iter.hasNext() ) {
	recurrence = iter.next().toInstant();
	recurrence = recurrence.truncatedTo( ChronoUnit.HOURS );
}

assertEquals( "2017-04-04T02:00:00.000Z", format.format( recurrence ) );
```

This fails as the date returned is 2017-04-27T00:00:00Z

If I print out the DateIterator values you can see that the advanceTo() feature has failed to bring the iterator to the right point. These are the dates it prints off (included only the first 36 occurrences). As you can see it starts from a date that's completely off, it looks like the advanceTo() feature doesn't take into account the days, only the year and month and then some how resets to the beginning of the month :-/

2017-04-27T00:00:00Z
2017-04-27T01:00:00Z
2017-04-27T02:00:00Z
2017-04-27T03:00:00Z
2017-04-27T04:00:00Z
2017-04-27T05:00:00Z
2017-04-27T06:00:00Z
2017-04-27T07:00:00Z
2017-04-27T08:00:00Z
2017-04-27T09:00:00Z
2017-04-27T10:00:00Z
2017-04-27T11:00:00Z
2017-04-27T12:00:00Z
2017-04-27T13:00:00Z
2017-04-27T14:00:00Z
2017-04-27T15:00:00Z
2017-04-27T16:00:00Z
2017-04-27T17:00:00Z
2017-04-27T18:00:00Z
2017-04-27T19:00:00Z
2017-04-27T20:00:00Z
2017-04-27T21:00:00Z
2017-04-27T22:00:00Z
2017-04-27T23:00:00Z
2017-04-01T00:00:00Z
2017-04-01T01:00:00Z
2017-04-01T02:00:00Z
2017-04-01T03:00:00Z
2017-04-01T04:00:00Z
2017-04-01T05:00:00Z
2017-04-01T06:00:00Z
2017-04-01T07:00:00Z
2017-04-01T08:00:00Z
2017-04-01T09:00:00Z
2017-04-01T10:00:00Z
2017-04-01T11:00:00Z
2017-04-01T12:00:00Z

**Comment by mangstadt** [comment]
Thanks for the report.  A number of other people have reported similar issues regarding the DateIterator.  Sorry you are having this problem.  Not sure what's causing these issues.

Issue may be related to #67.

**Comment by agiannone** [comment]
Hi there, so after a little digging I noticed that the advanceTo() method is only taking into account the year and month but skipping the day. That's why it advances to the correct year and month correctly but then the day is the same as the one in the periodStart instead of the awardedOn date.

**Comment by mangstadt** [comment]
Thanks for the insight.  To be honest, this date iteration functionality is very difficult for me to wrap my tiny little head around. :/

**Comment by mangstadt** [comment]
Fixed in 980e2da853f2248bbce47e52f1efdb1bc66820f3.

For what I assume to be performance reasons, the advanceTo method contains code that "jumps" the iterator ahead a certain amount, so that it does not have to iterate over every single date in the recurrence until it reaches the stop date. As you suggested, this code was not taking the day of the month into account. Setting the day of the month to 1 seems to solve the issue.

(This issue had nothing to do with https://github.com/mangstadt/biweekly/issues/67.)
